### DISCUSSION TITLE
Join kiali backend and frontend repos

### DISCUSSION THREAD

**Discussion by lucasponce** [kiali/kiali#4587] [General]
I really wonder if it's worth to move the kiali-ui repo as a "frontend" folder and unify it in a single repo, like other similar projects [1].

There is a strong justification when you need to provide different releases per repo, as it happens with the operator.

But in case of the backend/frontend both repos are part of the same Kiali releases and it may help to save some step.

[1] https://github.com/openshift/console/

**Comment by nrfox** [comment]
+1 for this. There's some overhead involved in linking UI and backend changes across repos as well as getting a development environment setup correctly across repos. Things like embedding the static content into the go binary would be easier if Kiali was a single repo.

**Comment by israel-hdez** [comment]
I think this can be subjective.
I would suggest to try first `git submodules` as a way to virtually try how it goes, before doing the actual move.

**Reply by jmazzitelli** [reply]
I don't like submodules. No one likes submodules :) It's just more git magic we have to worry about.

**Reply by israel-hdez** [reply]
It's not to keep them forever. It is a suggestion to try how it would be to have a single repo. If we see that it's good, we join the repos. Else, we get rid of the submodules :-)

**Reply by jmazzitelli** [reply]
I'm ok with it as a temporary "kick-the-tires, how would this work" kind of thing

**Reply by lucasponce** [reply]
Do you know if using submodules a single PR would affect both repos ? (I haven't tested it, but as @jmazzitelli says, that feature has not a lot of fans).

**Reply by israel-hdez** [reply]
AFAIK, you still need to open one PR on each repo :-/

**Reply by jmazzitelli** [reply]
I actually never looked into how PRs affect submodules. This is what I mean by "git magic" - https://stackoverflow.com/questions/47971173/pull-request-with-submodules

If I had a vote, I would say no to submodules. Too "weird" to me. And not many people know the tricks to using it, and it is just annoying.

**Reply by israel-hdez** [reply]
> I actually never looked into how PRs affect submodules. This is what I mean by "git magic" - https://stackoverflow.com/questions/47971173/pull-request-with-submodules

Oh, that would be a major annoyance.
Anyway, it was just a suggestion...

Also, as another suggestion, I would even include into consideration joining all repos. It's quite common that one issue needs to touch back-end, ui, operator and helm charts. Everything will be simplified, including CI, where you can run tests for the unified thing.

**Reply by lucasponce** [reply]
But I guess that operator would have a different version cadence, but I think that backend+frontend would always have same cadence.

Does it make sense to separate versions between UI and backend ? 

Also PRs can be better aligned as typically we say this is the PR for backend and this one for the UI, but I see it can get more clean to have it all together.

In the operator I see a different case, as I see it as a different artifact to deliver, I guess.

P.S.: No real problem with keep it split it, but for some time to time I compare with other projects and this comes to my mind as well. That's main motivation of the poll.

**Reply by jmazzitelli** [reply]
We have gone back and forth on this. Before, the operator and server had the same cadence/version. Then a couple years ago we split it. This enabled us to release the operator separate from the server (and that is how we are now). This means we can productized just the operator and re-used the server release. This saved time back when we needed to build RPMs and sign them. But now with CPaaS, we are releasing everything always - we cannot build one without the other (this isn't technically true for our older versions such as 1.12 and 1.24 but ignore that for now). So we don't really need the operator and server (and UI) all in separate repos for versioning. We could actually combine everything. We also have the helm-charts repo but I think we need to keep that separate because that is where we also publish the helm chart index.

I don't know if that makes things easier or not (combining operator with the server/ui repos). I guess we could create an issue for someone to research what would need to change and what the benefits and drawbacks would be.

**Comment by jshaughn** [comment]
This would affect our CPaaS, maybe minimally, I'm not sure.

**Reply by jmazzitelli** [reply]
I will affect that yes - but it would mean we just remove one of the mid-stream repos and simplify some of the config. It would actually make the CPaaS config less complex. Definitely would require changes.

**Reply by aljesusg** [reply]
From CPaaS perspective, we can do it, I saw the cachito docs etc..and we can have both in the same repo. 
And as @jmazzitelli said...this will make the config less complex :)

**Comment by jshaughn** [comment]
My gut feeling is that, if we were to combine repos, to leave the operator separate.  Just the fact that it operates on several releases makes it seem like a standalone repo makes sense, and combining it will have us always saying things like, "yes, but...".

As for the UI and Server, I'm ambivalent.  I like that the go code and javascript code are separated, and that the API is the exposure to client projects.  But it would be convenient to generate single PRs that span UI and server.  I don't think we need to release the UI and Server separately.  Having said that, in a dev env I certainly don't want to have to build each at the same time.

**Reply by jmazzitelli** [reply]
> the fact that it operates on several releases

That's a good point. Operator v1.36 supports 1.24 and 1.12 servers. I'm trying to think why we split it out in the first place, and I think one of the reasons is what you just brought up - the operator supports multiple versions of the server. So it isn't true that you should look at branch v1.12 to find out the operator that supports installing v1.12 --- the operator at the highest version will/should support that today (so you would look at v1.36 branch).

Thus, it would be wrong to say "I will build v1.12 to get the server and the operator that installs it." Building the v1.12 branch in the server repo will NOT build the correct operator.

**Comment by aljesusg** [comment]
IMO +1 on this, if we had a container for UI and another one from backend it would be another movie. In our case I think that we should join UI and backend in the same repo, It will take time and changes but in the long run everything will be easier for us as well as for QE.

**Reply by jmazzitelli** [reply]
Yes. I think combining the UI and the server will make building easier (and CPaaS config less complex).

I think we should keep the operator separate (I think it would be wrong for us to combine that too - this is due to the fact that the operator can support multiple versions of the server).
